roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+
+    // if we do not have the ball yet, go get it
+    double deltaPosSquared = (finalTargetPos - ball->pos).length2();
+    bool ballIsFarFromTarget = deltaPosSquared > 0.5;
+
+    bool robotDoesNotHaveBall = ! robot->hasBall();
+    bool robotIsTooFarFromBall = (robot->pos - ball->pos).length2() > maxBallDistance*maxBallDistance;
+    bool ballIsMovingTooFast = ball->vel.length2() > minVelForMovingball*minVelForMovingball;
+
+    if (robotDoesNotHaveBall && (robotIsTooFarFromBall || ballIsMovingTooFast)) {
+        return goToBall(ballIsFarFromTarget);
+    }
+
+    // check if we are doing something already
+    if (dribbleBackwards->getBackwardsProgression() != DribbleBackwards::START) {
+        return dribbleBackwards->getRobotCommand(robot, targetPos, targetAngle);
+    }
+
+    if (dribbleForwards->getForwardsProgression() != DribbleForwards::START) {
+        return dribbleForwards->getRobotCommand(robot, targetPos, targetAngle);
+    }
+
+    switch (preferredTravelStrategy) {
+    case forwards: return dribbleForwards->getRobotCommand(robot, targetPos, targetAngle);
+    case backwards: return dribbleBackwards->getRobotCommand(robot, targetPos, targetAngle);
+    case no_preference: {
+        // choose based on distance from the ball to the target
+        if (ballIsFarFromTarget) {
+            return dribbleForwards->getRobotCommand(robot, targetPos, targetAngle);
+        }
+        return dribbleBackwards->getRobotCommand(robot, targetPos, targetAngle);
+    }
+    }
+
+}
+
+RobotCommand BallHandlePosControl::goToBall(bool ballIsFarFromTarget, TravelStrategy preferredTravelStrategy) {
+
+    if (Constants::SHOW_BALL_HANDLE_DEBUG_INFO()) {
+        std::cout << "we do not have a ball yet" << std::endl;
+    }
+
+    if (ball->vel.length2() > minVelForMovingball*minVelForMovingball) {
+        ball->pos += (ball->vel*0.5 + ball->vel.stretchToLength(pow(ball->vel.length2(), 1.0/8.0)));
+    }
+
+    dribbleBackwards->reset();
+    dribbleForwards->reset();
+    RobotCommand robotCommand;
+    Vector2 target;
+    Vector2 ballToTarget = finalTargetPos - ball->pos;
+    if (preferredTravelStrategy == backwards) {
+        target = ball->pos + ballToTarget.stretchToLength(maxBallDistance);
+    }
+    else if (preferredTravelStrategy == forwards) {
+        target = ball->pos + ballToTarget.stretchToLength(- maxBallDistance);
+    }
+    else if (ballIsFarFromTarget) {
+        target = ball->pos + ballToTarget.stretchToLength(- maxBallDistance);
+    }
+    else {
+        target = ball->pos + ballToTarget.stretchToLength(maxBallDistance);
+    }
+
+    auto pva = numTreePosControl->getPosVelAngle(robot, target);
+    robotCommand.angle = (ball->pos - robot->pos).toAngle();
+    robotCommand.vel = pva.vel;
+    return robotCommand;
+}
+
+void BallHandlePosControl::setMaxVelocity(double maxV) {
+    maxForwardsVelocity = maxV > 0.0 && maxV < 8.0 ? maxV : maxForwardsVelocity;
```

Comment:
would be a bit cleaner if you make a boolean inVelocity range and use that in the ternary if statements, I think, since you are using it twice.

Reply:
fixed